Libplanet/Store/TrieStateStore.cs

```diff
@@ -90,6 +93,78 @@ public void ForkStates<T>(Guid sourceChainId, Guid destinationChainId, Block<T>
             // Do nothing.
         }
 
+        public void PruneStates(ImmutableHashSet<HashDigest<SHA256>> excludeBlockHashes)
+        {
+            var stopwatch = new Stopwatch();
+            Log.Debug($"Started {nameof(PruneStates)}()");
+            var excludeNodes = ImmutableHashSet<HashDigest<SHA256>>.Empty;
+            foreach (var blockHash in excludeBlockHashes)
+            {
+                if (!_stateHashKeyValueStore.Exists(blockHash.ToByteArray()))
+                {
+                    continue;
+                }
+
+                byte[] stateRootHashBytes = _stateHashKeyValueStore.Get(blockHash.ToByteArray());
+                var stateTrie = new MerkleTrie(
+                    _stateKeyValueStore,
+                    new HashNode(new HashDigest<SHA256>(stateRootHashBytes)));
+                Log.Debug("Start to iterate hash-nodes.");
+                stopwatch.Start();
+                var nodeHashes = stateTrie.IterateHashNodes();
+                excludeNodes = excludeNodes.Concat(nodeHashes).ToImmutableHashSet();
+                Log.Debug(
+                    $"Finished to iterate hash-nodes. elapsed: {stopwatch.ElapsedMilliseconds} ms");
+                stopwatch.Stop();
+            }
+
+            Log.Debug("ExcludedNodes' count: {Count}", excludeNodes.Count);
+
+            // Clean up nodes.
+            long deleteCount = 0;
+            Log.Debug("Started to clean up states.");
+            stopwatch.Restart();
+            foreach (var stateKey in _stateKeyValueStore.ListKeys())
+            {
+                if (excludeNodes.Contains(new HashDigest<SHA256>(stateKey)))
+                {
+                    continue;
+                }
+
+                _stateKeyValueStore.Delete(stateKey);
+                ++deleteCount;
+            }
+
+            Log.Debug(
+                "Finished to clean up states. Deleted {DeleteCount}." +
+                " elapsed: {ElapsedMilliseconds} ms",
+                deleteCount,
+                stopwatch.ElapsedMilliseconds);
+            stopwatch.Stop();
+
+            // Clean up state root hashes.
+            deleteCount = 0;
+            Log.Debug("Started to clean up state hashes.");
+            stopwatch.Restart();
+            foreach (var stateHashKey in _stateHashKeyValueStore.ListKeys())
+            {
+                if (excludeBlockHashes.Contains(new HashDigest<SHA256>(stateHashKey)))
```

Comment:
Maybe it's better for `.ListKeys()` to return `HashDigest<SHA256>` right away.

Reply:
You mean *change `byte[] IKeyValueStore.ListKeys()`'s signature into `HashDigest<SHA256> IKeyValueStore.ListKeys()`*? If then, I am not sure for the suggestion because it is not sure its key will be always `HashDigest<SHA256>`.